File: tests/Domain/Character/Health.test.ts
import { Health } from '@/Domain/Character/Health'
import { HitPointsDto } from '@/Domain/Character/HitPointsDto'
import { describe, test, expect } from 'vitest'

describe('Testing Health', () => {
  test('It should throw error when max below 0', () => {
    expect(() => {
      // eslint-disable-next-line
      Health.fromMaxHp(-1)
    }).toThrow(RangeError)
  })

  test('It should return proper values', () => {
    const sut = Health.fromMaxHp(12)
    const hitpoints = sut.hitpoints()
    expect(hitpoints).toBeInstanceOf(HitPointsDto)
    expect(hitpoints.max).equals(12)
    expect(hitpoints.current).equals(12)
  })
  test('It should return proper values when current less than max', () => {
    const sut = Health.fromValues(24, 12)
    const hitpoints = sut.hitpoints()
    expect(hitpoints.max).equals(24)
    expect(hitpoints.current).equals(12)
  })
})

File: backend/src/domain/character/health_test.go
package character

import (
	"testing"
)

func TestShouldThrowErrorWhenMaxHpZero(t *testing.T) {
	_, err := NewHealth(0)
	if err == nil {
		t.Errorf("Should throw error when max HP zero or less")
	}
}

func TestShouldReturnProperHitPointsValue(t *testing.T) {
	sut, _ := NewHealth(1)
	if sut.CurrentHp() != 1 {
		t.Errorf("Should return proper value of HitPoints")
	}
}

func TestShouldReturnProperCurrentHitpointsWhenDefined(t *testing.T) {
	sut, _ := NewHealthWithCurrent(10, 1)
	if sut.CurrentHp() != 1 {
		t.Errorf("Should return proper value of HitPoints")
	}
}

func TestShouldReturnTrueBleedingWhenBelowHalf(t *testing.T) {
	sut, _ := NewHealthWithCurrent(10, 5)
	if !sut.IsBleeding() {
		t.Errorf("Should return proper value of Bleeding")
	}
}

func TestShouldReturnFalseBleedingWhenAboveHalf(t *testing.T) {
	sut, _ := NewHealthWithCurrent(10, 6)
	if sut.IsBleeding() {
		t.Errorf("Should return proper value of Bleeding")
	}
}
